### DISCUSSION TITLE
Memory Upgrade

### DISCUSSION THREAD

**Discussion by csonyi** [P3pp3rF1y/SophisticatedBackpacks#93] [Ideas]
This is something that would be amazing if you could implement it.
What I have in mind is that when the upgrade is activated, every slot that has an item will save it (like the RfTools crafter), and when you shift click the item back in, it will go to its own slot.
When the item is not in its "remembered" slot, it could show a ghost image of it.

I usually use backpacks as kind of a toolbox, and the order of items changes all the time...
This would make looking for items that you use often (wrenches, niche tools, etc...) a breeze to find, since they could have a memorized spot that doesnt change after sorting the backpack.

**Comment by P3pp3rF1y** [comment]
Probably wouldn't make this into an upgrade, more like default functionality that will allow you to set slots with some modes in addition to saving an item and shift clicking first into there are two more I can think of:
- lock slot to prevent sorting (requested, could be the same or separate setting for slot)
- prevent automation from working with slots
And I had more in mind in the past just drawing a blank now.

What I am struggling with a bit is how this would look like in gui
- probably a button that would switch the gui between regular inventory view and "slots settings view" 
- and then in slots settings view you would have options displayed for each of the slots or maybe different colors for each of the modes set or small icons and hovering over slot could display more info / buttons to set these settings?

Any ideas are welcome

**Reply by csonyi** [reply]
I have two ideas in mind:
* As I mentioned in my post, the RFTools Crafters in 1.12 had a similar functionality. They had a button in the UI, and when you pressed it, any slot which had an item it saved that item as a ghost image, and from then on that slot only accepted that item. Shift+clicking the button could clear the saved slots.
* The other idea is either a button that when you press you enter a slot setting _mode_, where clicking a slot locks it to the item it has, clicking it again unlocks it, and again shift+clicking the button could clear all. The other way I would implement this is with a hotkey, that when pressed over a slot lock/unlocks it.

I understand not making it into an upgrade. When I thought about suggesting this here, the "Memory Upgrade" name popped into my head and I thought it sounded cool :D
This backpack mod is far superior to any other one I used in the past, Its honestly the best one so far.
Keep up the good work! ;)
